Fixed #27063 -- Prevented i18n_patterns() from using too much of the URL as the language.

urbaniak opened the pull request:
Fixing the issue described here https://code.djangoproject.com/ticket/27063.

That commit makes `get_language_from_path` returning only valid (defined in LANGUAGES setting) languages, so things like `/de-moines-parking/` is not a valid laguage anymore.

Before it assumes it's `de` language with `moines-parking` variant and returned `de`, just silently ignoring the invalid `monies-parking` part.

We should not assume that anything starting with a valid language code followed by a dash and any random string is a language with some random variant. That's changed only while looking up languages from path, the `Accept-Language` header parsing remains unchanged.

As looked at the code  I think we can also modify the regex to match only languages with one dash or at sign, AFAIK there's no case where language code contains two `@` or `-`.

Here's the proposed change to regex, not committed yet, cause I'm not 100% sure about it.

I'm waiting for feedback.

```
-language_code_prefix_re = re.compile(r'^/([\w@-]+)(/|$)')
+language_code_prefix_re = re.compile(r'^/(\w+([@-]\w+))(/|$)')
```

timgraham commented:
`(\w+([@-]\w+)?)` looks a bit more correct to me.
some characters, possibly followed by @- and some more characters

urbaniak commented:
That's right, I've forgot about the question mark

urbaniak commented:
Tried the approach with changing the regexp, it's working fine for me.

Added also the regression tests.

claudep commented:
I'd like to see some more targeted tests in `test_get_language_from_path_real`.
Please merge your commits and update the commit message.

urbaniak commented:
Squashed the commits and added some more tests in `test_get_language_from_path_real`.

claudep commented:
Thanks! @timgraham, does it look ready for you?

timgraham commented on tests/i18n/tests.py:
```diff
@@ -1472,12 +1472,32 @@ def test_parse_language_cookie(self):
+        self.assertEqual(g('/de-at/'), 'de-at')
+        self.assertEqual(g('/de-ch/'), 'de')
+
+        # Regresion test for #27063
```
Describe the issue in a comment rather than including a ticket reference (saves a lookup step to understand the purpose of the test).

timgraham commented:
To better clarify what's fixed, any tests that pass before this change should be included in a separate commit.

I suppose this should be backported to 1.10 since `prefix_default_language` is new there. Can you add a release notes to docs/releases/1.10.2.txt?

urbaniak commented:
I've split those changes into two commits - one confirming the issue and another fixing it.

Added the release notes.

timgraham commented on tests/i18n/tests.py:
```diff
@@ -1472,12 +1472,32 @@ def test_parse_language_cookie(self):
+        self.assertEqual(g('/de-ch/'), 'de')
+
+        # Regresion #27063: Make sure that pages starting with /de* dont change language to german
+        self.assertIsNone(g('/de-simple-page/'))
```
What I meant about splitting the tests in a separate commit only applies to the tests that already pass before the fix. The tests that are fixed by the change should go in the second commit along with the fix.
Actually, I only see this assertion failing before the fix, so it might be that the other tests need adjusting so that they're proper regression tests: `self.assertIsNone(g('/de-simple-page/'))`

timgraham commented on docs/releases/1.10.2.txt:
```diff
@@ -14,3 +14,6 @@ Bugfixes
 * Allowed combining ``contrib.postgres.search.SearchQuery`` with more than one
   ``&`` or ``|`` operators (:ticket:`27143`).
+
+* Fixed issue where pages starting with /de-* are matched to german language.
```
Of course this applies to any language that has similar matching, so try to make the note a bit more general.

urbaniak commented:
Sorry for a delay.

I've updated the PR, added release notes to 1.10.3.txt.

timgraham commented on tests/i18n/tests.py:
```diff
@@ -1854,6 +1859,15 @@ def test_unexpected_kwarg_to_i18n_patterns(self):
         with self.assertRaisesMessage(AssertionError, "Unexpected kwargs for i18n_patterns(): {'foo':"):
             i18n_patterns(object(), foo='bar')
 
+    # Regresion #27063: Make sure that pages starting with /de* dont change language to german
```
These two tests are still passing even with the fix reverted.

urbaniak commented:
Removed those two not needed tests.

urbaniak commented:
Removed one of the tests cause it was really not failing, the second was not failing because 'de' was not in the overridden LANGUAGES, which I've added now.

timgraham commented:
merged in ee06689878e87e45d3c2402dd5e3bb3ba2a79bcf and a01d887a3ad9029ed198b6af974534fec46223f9, thanks!
